src/broadcaster.rs

```diff
@@ -0,0 +1,192 @@
+            targets: Mutex::new(vec![Arc::downgrade(&status)])
+        });
+
+        let shared_state = Arc::new(Mutex::new(BroadcasterSharedState {
+            signal: signal,
+            value: None,
+            notifier: notifier
+        }));
+
+        let state = Arc::new(BroadcasterState {
+            status: status,
+            shared_state: shared_state
+        });
+
+        Self {
+            state: state
+        }
+    }
+}
+
+impl<A: Signal> Signal for Broadcaster<A>
+    where <A as Signal>::Item: Clone  {
+    type Item = <A as Signal>::Item;
+
+    fn poll_change(&mut self, cx: &mut Context) -> Async<Option<Self::Item>> {
+        // NB we musn't hold the lock on self.state.status during the call to
+        // shared_state.poll_change(), as that must access all the children to notify them.
+
+        // Don't need any potential waker as we're already polling
+        self.state.status.lock().unwrap().waker = None;
+
+        let mut shared_state = self.state.shared_state.lock().unwrap();
+        if let Async::Ready(value) = shared_state.poll_change(cx) {
+            // value was just broadcast
+            self.state.status.lock().unwrap().has_changed = false;
+            Async::Ready(value)
+        } else {
+            let mut status = self.state.status.lock().unwrap();
+            if status.has_changed {
+                // a new value was polled and broadcast by another task since we last polled
+                status.has_changed = false;
+                Async::Ready(shared_state.value.clone())
+            } else {
+                // nothing new to report
+                status.waker = Some(cx.waker().clone());
+                Async::Pending
+            }
+        }
+    }
+}
+
+impl<A: Signal> Clone for Broadcaster<A>
+    where <A as Signal>::Item: Clone {
+    fn clone(&self) -> Self {
+
+        let new_status = Arc::new(Mutex::new(BroadcasterStatus {
+            has_changed: self.state.status.lock().unwrap().has_changed,
+            waker: None
+        }));
+
+        let shared_state = self.state.shared_state.lock().unwrap();
+        shared_state.notifier.targets.lock().unwrap().push(Arc::downgrade(&new_status));
+
+        let new_state = Arc::new(BroadcasterState {
+            status: new_status,
+            shared_state: self.state.shared_state.clone()
+        });
+
+        Self {
+            state: new_state,
+        }
+    }
+}
+
+// ----------------------------------------------------------------------------------------------
+
+// State for a single broadcaster instance.
+//
+// It's split in this way because the BroadcasterSharedState also needs to access the status
+// (e.g. to notify of changes).
+struct BroadcasterState<A: Signal> {
+    status: Arc<Mutex<BroadcasterStatus>>,
+    shared_state: Arc<Mutex<BroadcasterSharedState<A>>>,
+}
+
+// ---------------------------------------------------------------------------------------------
+
+struct BroadcasterStatus {
```

Comment:
Oh, also, this should probably be like this:

```rust
struct BroadcasterStatus {
    has_changed: AtomicBool,
    waker: Mutex<Option<Waker>>,
}
```

Reply:
And then it only needs to use `Arc<BroadcasterStatus>`

Reply:
Yes I was really torn about which way round to put the `mutex`. In the end I went for outside the struct because then the struct itself is simpler, but I guess with the mutex on the field chance of a deadlock is reduced. Will change.

Reply:
The personal guideline I use is that if every field except for one can use the `Atomic*` types, then I use the `Atomic*` types + 1 `Mutex` in the struct. Otherwise I put the `Mutex` outside the struct.

Of course there's a lot of exceptions to that: situations where you need to lock two fields separately, situations where you would prefer to use `RwLock` instead, etc.

Reply:
Oh, also, I prefer using `RwLock` if some parts of the code need mutable access and some parts need read-only access.

But if every part needs mutable access then I use `Mutex` (since it's lighter-weight than `RwLock`)